radix-engine/src/vm/wasm/wasmi.rs

```diff
@@ -957,17 +1001,17 @@ impl WasmEngine for WasmiEngine {
             #[cfg(not(feature = "moka"))]
             {
                 if let Some(cached_module) = self.modules_cache.borrow_mut().get(metered_code_key) {
-                    return cached_module.instantiate();
+                    return Ok(cached_module.instantiate());
                 }
             }
             #[cfg(feature = "moka")]
             if let Some(cached_module) = self.modules_cache.get(metered_code_key) {
-                return cached_module.as_ref().instantiate();
+                return Ok(cached_module.as_ref().instantiate());
             }
         }
 
         let code = &instrumented_code.code.as_ref()[..];
-        let module = WasmiModule::new(code).unwrap();
+        let module = WasmiModule::new(code)?;
```

Comment:
I'm not sure about the error propagation here as this not failing I think is an invariant of the system and we don't currently distinguish between invariant-breaking errors vs. non-invariant-breaking errors.

An example of how this may affect things practically: if a validator has some misconfiguration or broken hardware which causes this error to occur they would rather panic rather than vote on a different result on a transaction from the other validators.

Reply:
Hmm so I would argue that there's nothing really wrong with a single validator misagreeing and voting for a different outcome. They still wouldn't be able to sync past this transaction, they would still end up panicking later when they try to ingest the ledger from ledger sync and get a byzantine quorum exception (or similar) when they get a different merkle hash.

Conversely - consider if there's a bug in Wasmi - or we upgrade the instrumenter - such that after an instrumentation change,  the instrumented code in a particular package couldn't be run any more - and this wasn't caught at testing time.

Would you rather a liveness break? Or validators agree that the transaction failed?

Personally I'd prefer the latter.

Reply:
Or if we really want to panic, we should at least be panicking higher up the stack in `radix-engine/src/vm/vm.rs` rather than here :D

Reply:
The concern is not necessarily a liveness break but as a validator I would rather panic than vote incorrectly and get slashed.

But yes panic prolly makes more sense at a higher layer where the invariant is defined.

Reply:
I'm not sure if voting incorrectly is slashable? Maybe it is. I can add the panic back in `radix-engine/src/vm/vm.rs` if we want.

More generally perhaps in future we need to have "invariant broken" kinds of errors which would let us return a different transaction result (and move the choice to panic or not back into the node) -- very similar to how we have the Aborted transaction result.